diff --git a/bundle/hjson.js b/bundle/hjson.js
index 8c260d7b71fbe81f3456cd513aecd46a2060ef10..d0f6ff67e628fbc74624841252c7de5efb740a92 100644
--- a/bundle/hjson.js
+++ b/bundle/hjson.js
@@ -218,8 +218,6 @@
 /* Hjson https://hjson.github.io */
   "use strict";
 
-  var os=require('os'); // will be {} when used in a browser
-
   function tryParseNumber(text, stopAtNext) {
 
   // try to parse a number
@@ -316,7 +314,7 @@
   }
 
   module.exports = {
-    EOL: os.EOL || '\n',
+    EOL: '\n',
     tryParseNumber: tryParseNumber,
     createComment: createComment,
     removeComment: removeComment,
diff --git a/lib/hjson-common.js b/lib/hjson-common.js
index 50546e85da7e2bef9952549b135776036d80683d..8d9d4f598a2c2df7de2ab7ceb3286b236a3e9dd5 100644
--- a/lib/hjson-common.js
+++ b/lib/hjson-common.js
@@ -1,8 +1,6 @@
 /* Hjson https://hjson.github.io */
 "use strict";
 
-var os=require('os'); // will be {} when used in a browser
-
 function tryParseNumber(text, stopAtNext) {
 
   // try to parse a number
@@ -99,7 +97,7 @@ function forceComment(text) {
 }
 
 module.exports = {
-  EOL: os.EOL || '\n',
+  EOL: '\n',
   tryParseNumber: tryParseNumber,
   createComment: createComment,
   removeComment: removeComment,
